test/compress/collapse_vars.js

```diff
@@ -2256,3 +2256,31 @@ issue_2187_3: {
     }
     expect_stdout: "1"
 }
+
+issue_2203_1: {
+    options = {
+        collapse_vars: true,
+        unused: true,
+    }
+    input: {
+        console.log({
+            a: 42,
+            b: function() {
+                return function(c) {
+                    return c.a;
+                }(this);
```

Comment:
One-upped your suggestion and did this:
https://github.com/mishoo/UglifyJS2/pull/2204/files#diff-37dcf1d17b8a2b3c980e12edef17fdc8R2290